cdap-api/src/main/java/io/cdap/cdap/api/app/Application.java

```diff
@@ -32,4 +32,17 @@
    * @param context Used to access the environment, application configuration, and application (deployment) arguments
    */
   void configure(ApplicationConfigurer configurer, ApplicationContext<T> context);
+
+  /**
+   * Returns if application supports config upgrade or not.
+   */
+  boolean allowConfigUpgrade();
+
+  /**
+   * Upgrades given configuration.
+   *
+   * @param config Configuration, most likely older, to be upgraded.
+   * @param applicationUpgradeContext Used to access methods helpful for upgrading a config such as get latest plugin versions.
+   */
+  ConfigUpgradeResult<T> upgradeConfig(T config, ApplicationUpgradeContext applicationUpgradeContext) throws Exception;
```

Comment:
should have a default implementation here so that existing apps don't need to be updated.

The default can just return whatever it got as input, and return empty results for upgraded artifacts.

Reply:
Returning same as input would just create confusion for developer and users right? IMO it is better to say this is not supported at all and its also easy to show during dry run that "Hey we can not upgrade your xyz pipeline in abc application cause it is not supported". wdyt?

Reply:
isn't that what the other boolean method is for?